PSReadLine/PlatformWindows.cs

```diff
@@ -614,4 +611,350 @@ public override void BlankRestOfLine()
+
+        if (diff < 0)
+        {
+            diff = 0;
+        }
+        else if (diff >= (long) Int32.MaxValue)
+        {
+            // Should not ever actually happen...
+            diff = DefaultGraceMillis;
+        }
+
+        return (int) diff;
+    }
+
+    private const int MaxRounds = 2;
+    private const bool ClearProgress = true;
+
+    //
+    //                         TerminateOrphanedConsoleApps
+    //
+    // This feature works around a bad interaction on Windows between:
+    //    * a race condition between ctrl+c and console attachment, and
+    //    * poor behavior when multiple processes want console input.
+    //
+    // This bad interaction is most likely to happen when the user has launched a process
+    // that is launching many, MANY more child processes (imagine a build system, for
+    // example): if the user types ctrl+c to cancel, all processes *currently attached* to
+    // the console will receive the ctrl+c signal (and presumably exit). However, there
+    // *may* have been some processes that had been created, but are not yet attached to
+    // the console--these grandchildren will have missed the ctrl+c signal (that's the
+    // race condition). If those grandchildren do not somehow figure out on their own that
+    // they should exit, the console enters a highly problematic state ("the borked
+    // state"): because pwsh's immediate child has exited, the shell will return to the
+    // prompt and wait for input. But those straggler granchildren are ALSO attached to
+    // the console... so when the user starts typing, who gets the input?
+    //
+    // It turns out that the console will just sort of randomly distribute pieces of input
+    // between all processes who want input--a straggler grandchild process might get a
+    // "key down" record, and then PSReadLine might get the corresponding "key up". This
+    // is obviously untenable; it makes the shell totally unusable. (The console team has
+    // been made aware, and there are several ideas of how to Do Better, but who knows
+    // when any of those will come to fruition.)
+    //
+    // To make matters worse: when returning to the prompt, PSReadLine disables ctrl+c
+    // signals (we prefer to handle those keys specially ourselves). So if you hit this
+    // situation with cmd.exe as your shell, you can just mash on ctrl+c for a while and
+    // kill all the stragglers manually; but if you have PSReadLine loaded, your shell is
+    // borked, and you are stuck. You CAN recover, IF you can track down and kill all the
+    // straggler processes manually.
+    //
+    // So when enabled, this feature does that for you: it kills all those straggler
+    // processes, right before we disable ctrl+c signals and wait for user input, ensuring
+    // that the user has a usable shell.
+    //
+    // Note that GUI processes do not attach to the console, so if you have launched
```

Comment:
I need to build the changes and see what it does when PowerShell is working with child process with stdin redirected. Haven't got the chance to do so, but will reply back if I find any issue with that.

Reply:
I played with the changes and found an issue:

> $p = Start-Process pwsh -ArgumentList '-c Write-Host start $pid; sleep -seconds 30; Write-Host stop' -NoNewWindow -passThru

After the option is enabled, when starting a console application like above, the process will be terminated immediately even though it doesn't really consume console input.

I don't think this is the right behavior. In this case, the new process shouldn't be considered as an "orphaned process" from the user's viewpoint.

Maybe we should check if a console-attached process is actually an orphan by verifying if its parent process id still points to an alive process that was created earlier.

Reply:
I played with the changes and found an issue:

> $p = Start-Process pwsh -ArgumentList '-c Write-Host start $pid; sleep -seconds 30; Write-Host stop' -NoNewWindow -passThru

After the option is enabled, when starting a console application like above, the process will be terminated immediately even though it doesn't really consume console input.

I don't think this is the right behavior. In this case, the new process shouldn't be considered as an "orphaned process" from the user's viewpoint.

Maybe we should check if a console-attached process is actually an orphan by verifying if its parent process id still points to an alive process that was created earlier.

Reply:
Actually, this is *exactly* the sort of process that we want to kill. Although you came up with a *mostly* benign example (it only spews a few pieces of "junk" into the middle of your interactive session), this process is very much capable of hosing your shell (more than just writing output while you are at the prompt).

Here is a modified example, which demonstrates that it is connected to the console's input (even without PSReadLine loaded, it will have weird behavior):

`$p = Start-Process pwsh -ArgumentList '-NoProfile -c Write-Host start $pid; write-host "input redirected? $([Console]::IsInputRedirected)" ; write-host "output redirected? $([Console]::IsOutputRedirected)" ; Write-Host "give me some input: " ; $userInput = [Console]::ReadLine() ; Write-Host "input was: $userInput" ;  sleep -seconds 30; Write-Host stop' -NoNewWindow -passThru`

Maybe you were thinking "well, it *could* read input, but if it *doesn't*, then it should be allowed". Unfortunately there is (of course) no way to tell if an arbitrary process is *going* to read input or not. But the fact that it *can* (while we are also waiting for input) qualifies it for termination.

There are some mitigating factors here:

1. (The biggest, IMO): it seems pretty unlikely that someone would do something like this directly from the command line. If they do it in a script, it is much more likely that the script is doing something sensible with the child process, and not returning to the prompt "early" (before the child process has exited), in which case everything would be fine.
2. This new option is off by default, so it definitely won't start causing a bunch of people to randomly have their processes killed.

Also, I think this aligns well with documented guidance for how/when to use `Start-Process` ([https://github.com/MicrosoftDocs/PowerShell-Docs/issues/6239](https://github.com/MicrosoftDocs/PowerShell-Docs/issues/6239)).

Reply:
> Maybe you were thinking "well, it could read input, but if it doesn't, then it should be allowed". Unfortunately there is (of course) no way to tell if an arbitrary process is going to read input or not. But the fact that it can (while we are also waiting for input) qualifies it for termination.

I think the criteria should be whether a console-attached process is an orphan.

It's OK if a user explicitly starts a process from PowerShell that also consumes console input -- they are allowed to shoot their own foot. But we should not prevent a user from do so to just run some background work.

The issue we want to address is to terminate the orphan processes spawned by a build process that was terminated, so we should make sure the changes doesn't go beyond that and breaking a legit scenario.
- for the target issue we want to address, it's desired for the user to not know a thing about PSReadLine killing those orphan processes, because they should have been gone with the user press `ctrl+c` to terminate the _build_ parent process.
- for the side effect that cause `start-process` to not work as expected, it will result in lots of troubles for the user to diagnose why the process was killed.

Reply:
Okay, my personal feeling is "I wouldn't like that"... but I understand that it could be difficult for someone to figure out why such a process was killed; and given that it should not be a problem for the scenario I am trying to solve, I am willing to compromise :D so I added the check.

Reply:
Thank you. I will do another review soon (first thing tomorrow morning if cannot be earlier :))

Reply:
Immediate children may have children processes that accidentally derives the standard input (which I agree is a bad thing to do), so ideally we should check the parent id of the console-attached process to make sure it still points to an alive process that was created earlier.

But I think we can just check for immediate children for simplicity, and wait for feedback.